engine/crates/engine-v2/schema/src/wrapping.rs

```diff
@@ -0,0 +1,272 @@
+    ///
+    /// - `String` => false
+    /// - `String!` => true
+    /// - `[String!]` => true
+    /// - `[String]!` => false
+    pub fn inner_is_required(self) -> bool {
+        self.0 & INNER_IS_REQUIRED_FLAG != 0
+    }
+
+    /// Innermost to outermost.
+    pub fn list_wrappings(self) -> impl DoubleEndedIterator<Item = ListWrapping> + Copy + Clone {
+        self
+    }
+
+    pub fn is_required(self) -> bool {
+        self.list_wrappings()
+            .next_back()
+            .map(|lw| matches!(lw, ListWrapping::RequiredList))
+            .unwrap_or(self.inner_is_required())
+    }
+
+    pub fn is_list(self) -> bool {
+        self.list_wrappings().next().is_some()
+    }
+
+    fn start(self) -> u32 {
+        (self.0 & START_MASK) >> START_SHIFT
+    }
+
+    fn inc_start(&mut self) {
+        let start = self.start() + 1;
+        self.0 = (self.0 & !START_MASK) | (start << START_SHIFT);
+    }
+
+    fn end(self) -> u32 {
+        (self.0 & END_MASK) >> END_SHIFT
+    }
+    fn inc_end(&mut self) {
+        let end = self.end() + 1;
+        assert!(end < MAX_LIST_WRAPINGS + 1, "Too many list wrappings");
+        self.0 = (self.0 & !END_MASK) | (end << END_SHIFT);
+    }
+
+    fn dec_end(&mut self) {
+        let end = self.end() - 1;
+        self.0 = (self.0 & !END_MASK) | (end << END_SHIFT);
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ListWrapping {
+    RequiredList,
+    NullableList,
+}
+
+impl Iterator for Wrapping {
+    type Item = ListWrapping;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let start = self.start();
+        if start == self.end() {
+            return None;
+        }
+        self.inc_start();
+        let bit_mask = 1 << start;
+        if self.0 & bit_mask != 0 {
+            Some(ListWrapping::RequiredList)
+        } else {
+            Some(ListWrapping::NullableList)
+        }
+    }
+}
+
+impl DoubleEndedIterator for Wrapping {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        let end = self.end();
+        if end == self.start() {
+            return None;
+        }
+        self.dec_end();
+        // end is exclusive
+        let bit_mask = 1 << (end - 1);
+        if self.0 & bit_mask != 0 {
+            Some(ListWrapping::RequiredList)
+        } else {
+            Some(ListWrapping::NullableList)
+        }
+    }
+}
+
+impl Wrapping {
+    pub fn new(required: bool) -> Self {
+        if required {
+            Self::required()
+        } else {
+            Self::nullable()
+        }
+    }
+
+    pub fn nullable() -> Self {
+        Wrapping(0)
+    }
+
+    pub fn required() -> Self {
+        Wrapping(INNER_IS_REQUIRED_FLAG)
+    }
+
+    #[must_use]
+    pub fn wrapped_by_nullable_list(mut self) -> Self {
+        let end = self.end();
+        self.inc_end();
+        let bit_mask = 1 << end;
+        self.0 &= !bit_mask;
+        self
+    }
+
+    #[must_use]
+    pub fn wrapped_by_required_list(mut self) -> Self {
```

Comment:
I like this function name, will probably steal it.

Reply:
haha feel free, stole tons of stuff from you :D